### PR TITLE
Allow .xcworkspace and .xcodeproj to be renamed from default name 'Runner'

### LINKED ISSUE #9767: Would like to be able to change ios/Runner.xcodeproj to ios/MyName.xcodeproj
@lukef brought up the question this morning if it would be possible to change ios/Runner.xcodeproj to ios/MyName.xcodeproj.  My understanding from @cbracken is that there are several places in our code which currently look for the word "Runner" or assume that path which would break if that was done.

We could certainly support this, but my understanding is changing the name is not "supported" out of the box.  I don't know what the situation is on the Android side (if there are any path naming issues there).

### PR DESCRIPTION
**Pull request by LouiseHsu**
Adds the ability to rename Runner.xcodeproj and Runner.xcworkspace - fixes https://github.com/flutter/flutter/issues/9767.

To rename a project:
1. Open Runner.xcodeproj in Xcode
2. In the left panel, left click "Show File Inspector" 
Screenshot 2023-04-17 at 11 41 07 PM
3. In the right panel, the name of the project, "Runner", should be visible under "Identity and Type". Change the name and press enter.
Screenshot 2023-04-17 at 11 40 43 PM
4. A wizard should pop up. Click Rename.
Screenshot 2023-04-17 at 11 44 01 PM
To rename the workspace:

1. Make sure Xcode is closed.
2. Rename the .xcworkspace to your new name.

If you also renamed the project

&nbsp; 3. Reopen the .xcworkspace in Xcode. If the selected project is the old name and in red, update it to match the new project name.

Tests for schemeFor were changed as with Xcode 14, in some cases the scheme will be renamed along with the project. Thus we will get the best match scheme for either the project name, or the default name Runner. However if a flavor is present, the scheme should always match the flavor.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:314]: Isn't hostAppProjectName a property on this instance?
```suggestion
      scheme = info.schemeFor(flavor: buildInfo?.flavor, hostAppProjectName: hostAppProjectName);
```

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:630]: What happened to this test?

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:663]: I don't see a test that's passing in a flavor and hostAppProjectName.  Imagine you had a bug that preferred the app name over the flavor.

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/xcodeproj.dart:460]: Was this change on purpose?

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (LouiseHsu) [reply packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:630]: this test was removed because the helper function it tests (expectedSchemeFor) was removed.

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (LouiseHsu) [reply packages/flutter_tools/lib/src/xcode_project.dart:43]: i removed the try catch - originally i added it because the cocoapod tests was throwing an exception because it tried to grab the hostAppProjectName with an empty test directory. I just added the corresponding macos and ios directories to the test directory, so it should be fine now.

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart
Comment (LouiseHsu) [reply packages/flutter_tools/lib/src/ios/xcodeproj.dart:486]: i actually think it should be mandatory? because this function needs to know what the project is called (either Runner or a custom name) and anything calling this function wouldn't be aware if the project was renamed.

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [reply packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:630]: I more meant I don't see an equivalent test for your `schemeFor` that is checking it's using title case.

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:648]: This line is getting log, suggest adding a comma after `logger` and putting each on its own line.
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line

File: packages/flutter_tools/test/general.shard/macos/cocoapods_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/macos/cocoapods_test.dart:64]: Since it threw without this, should `defaultHostAppProjectName` instead check if the project exists first if that codepath is possible?

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/xcodeproj.dart:480]: Suggest keeping `hostAppProjectName` as a named parameter as you had before, but mark it required:
```suggestion
  String? schemeFor({required String hostAppProjectName, String? flavor}) {
```

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:46]: Instead of caching to `_hostAppProjectName` you could use `late final` which will only run once:

```suggestion
  late final String hostAppProjectName = () {
    const String defaultHostAppProjectName = 'Runner';
    List<FileSystemEntity> contents;
      if (!hostAppRoot.existsSync()) {
        return defaultHostAppProjectName;
      }
      contents = hostAppRoot.listSync();
      final FileSystem fileSystem = hostAppRoot.fileSystem;
      for (final Directory entity in contents.whereType<Directory>()) {
        // On certain volume types, there is sometimes a stray `._Runner.xcworkspace` file.
        // Find the first non-hidden xcworkspace and return the directory.
        if (fileSystem.path.extension(entity.path) == '.xcworkspace' &&
            !fileSystem.path.basename(entity.path).startsWith('.')) {
          return fileSystem.path.basenameWithoutExtension(entity.path);
        }
      }
      return defaultHostAppProjectName;
  }();
```
`defaultHostAppProjectName` is only used in this method so can be made a const local.
Also you can grab the filesystem off `hostAppRoot` so you don't need to use the `globals` context.

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:623]: `BuildInfo.debug.flavor` is null in all these cases, right? It's not set?  So these are all the same:

```dart
    expect(info.schemeFor(hostAppProjectName: 'Runner'), 'Runner');
    expect(info.schemeFor(hostAppProjectName: 'Runner'), 'Runner');
    expect(info.schemeFor(hostAppProjectName: 'Runner'), 'Runner');
```

Seems like a good place to check casing:

```suggestion
    expect(info.schemeFor(hostAppProjectName: 'Runner'), 'Runner');
    expect(info.schemeFor(hostAppProjectName: 'runner'), 'Runner');
    expect(info.schemeFor(hostAppProjectName: 'RUNNER'), 'Runner');
```

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:651]: Again I think these are all the same because `BuildInfo.*.flavor` is null
```suggestion
    expect(info.schemeFor(hostAppProjectName: 'Runner'), 'Runner');
```

File: packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart:685]:
```suggestion
    expect(info.schemeFor(hostAppProjectName: 'Runner'), isNull);
```

**Comment by jmagman** [comment]
@LouiseHsu I should stop reviewing this until you mark this as not a draft and ready for review, sorry about that 🙂

**Comment by LouiseHsu** [comment]
@jmagman my bad, I totally forgot to put this into review mode. Thank you for the reviews!

**Comment by jmagman** [comment]
@LouiseHsu Can you fill out the description with the details of this change? Why did you have to touch schemes, how to test it, for example.  

Additionally "Allow runner rename" is not a very descriptive commit title since anyone unfamiliar with the problem will have any idea what it means. Something like "Allow iOS and macOS Xcode project to be renamable from default Runner" would be better

You can also include "Fixes https://github.com/flutter/flutter/issues/9767" and the "fixes" part will close that issue when this PR closes.

**Comment by hellohuanlin** [comment]
> 3. In the right panel, the name of the project, "Runner", should be visible under "Identity and Type". Change the name and press enter.

Can also just select the project on the navigation tab on the left, hit enter, and type the new name.

**Comment by hellohuanlin** [comment]
> Tests for schemeFor were changed as with Xcode 14, in some cases the scheme will be renamed along with the project. 

I think we can also instruct people to rename the scheme. Or ask them to rebuild the iOS project, which should update to the new scheme name.

**Comment by hellohuanlin** [comment]
I renamed a bunch of my personal projects before. I think we should also rename the folder `ios/Runner/class_files`, and also update `Info.plist` path after that (but maybe flutter build already have that step covered)

**Review by jmagman — CHANGES_REQUESTED** [review]
> To also rename the workspace:
> 1. Close Xcode after steps 1 to 4.
> 2. Rename the .xcworkspace to your new name.

Can you make this PR work without requiring the xcworkspace to be renamed?  Or just the workspace renamed but not the project?  I updated just my project and not the workspace and got:

```
Expected ios/Runner.xcodeproj but this file is missing.
Application not configured for iOS
```

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:49]: This directory is still `Runner` though, right? That wouldn't necessarily be changed.  When I tested a `xcodeproj`  rename it was still at `Runner`.

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:518]: This doesn't seem right either, before I renamed the Xcode project these files were at `ios/Runner/GeneratedPluginRegistrant.h` but on this PR after the rename it started generating it in `ios/Runner1` which isn't used anywhere.

**Review by jmagman — CHANGES_REQUESTED** [review]

**Comment by jmagman** [comment]
> I think we can also instruct people to rename the scheme. Or ask them to rebuild the iOS project, which should update to the new scheme name.

> I renamed a bunch of my personal projects before. I think we should also rename the folder `ios/Runner/class_files`, and also update `Info.plist` path after that (but maybe flutter build already have that step covered)

I think the point of this issue is that we need to handle any combination of renames that Xcode allows, not assume the name of the project or workspace.  If it works from Xcode it should work from `flutter`.  Users should be allowed to just rename the project or workspace, or name them different things, as they choose.  The `Runner` directory is trickier, but we can figure that out in another PR.

**Comment by jmagman** [comment]
In fact I think these should be addressed separately, if at all. 
1. The xcodeproj or xcworkspace is renamed, maybe not the same.  This is trivial to detect by the extension and was the only thing I considered in scope to this PR.
2. The scheme is also renamed.  We already handle schemes with flavors and I would wait until we get any feedback from users that they want to rename their default project scheme without using flavors.
3. They want to rename their `ios/Runner` directory.  This would would be much trickier since `flutter` generates files to that directory and is opinionated about what's in there.   I don't think I've seen any case of a user requesting that and if they did it sounds like a P6.

Let's narrow the scope of this to 1: The xcodeproj or xcworkspace is renamed.  That's it, not the scheme, not the directory.

**Comment by jmagman** [comment]
> Additionally "Allow runner rename" is not a very descriptive commit title since anyone unfamiliar with the problem will have no idea what it means. Something like "Allow iOS and macOS Xcode project to be renamable from default Runner" would be better
> 
> You can also include "Fixes #9767" and the "fixes" part will close that issue when this PR closes.

Bump these two comments as well 🙂

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart
Comment (vashworth) [inline packages/flutter_tools/lib/src/ios/xcodeproj.dart:495]: nit: Doesn't seem like `matchingScheme` needs to be declared before the if and looks like it can be final.
```dart
if (schemes.contains(expectedScheme)) {
  return expectedScheme;
}
final String? matchingScheme = _uniqueMatch(schemes, (String candidate) {
```

**Review by vashworth — COMMENTED** [review]
I 👍 some of @jmagman's comments that still need to be addressed

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/xcodeproj.dart:463]: Seems like you don't need to rearrange this method location?

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:24]:
```suggestion
  final String _defaultHostAppProjectName = 'Runner';
```

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:48]: Maybe this shouldn't be called `defaultHostAppProjectName` if it's actually the directory, and it's not related to the `hostAppProjectName` logic.

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:517]: Same comment about the directory name.

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:285]: I changed my project and workspace names and it still builds Runner.app.

This wouldn't be changed by the Xcode project name, the intent here was to fall back to `Runner.app` which would be the default from the `flutter create` template:

https://github.com/flutter/flutter/blob/635337ca1556b7a6cb50bb9ca0d36721e9772234/packages/flutter_tools/templates/app_shared/ios-objc.tmpl/Runner.xcodeproj/project.pbxproj.tmpl#L390

(`TARGET_NAME` is Runner)

Screenshot 2023-05-01 at 4 29 26 PM

So this should be treated more like the directory name.  Though at that point `_defaultHostAppProjectDirectoryName` isn't a good name since it's representing the fallback for project name, directory name, and product name.

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:24]: Prefer const over final if the value is known at compilation time.
```suggestion
  static const String _defaultHostAppName = 'Runner';
```

**Review by jmagman — APPROVED** [review]
LGTM after minor `const` nit changed.

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [reply packages/flutter_tools/lib/src/xcode_project.dart]: If it's `static` that means it's on the class, not instance, you need to call as `XcodeBasedProject._defaultHostAppName`.  I didn't include that suggestion but this by itself won't compile.

**Comment by vincekruger** [comment]
Testing out this change with the latest version of Flutter.  Initially, without any additional pods installed, the name change functions correctly, however, once adding `firebase_core` a pod file is generated and the build fails.

```
Launching lib/main.dart on iPhone 14 Pro Max in debug mode...
main.dart:1
CocoaPods' output:
↳
      Preparing
    Analyzing dependencies
    Inspecting targets to integrate
    [!] Unable to find the Xcode project `/{pathtoproject}/ios/Runner.xcodeproj` for the target `Pods-Runner`.
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer/analyzer/target_inspector.rb:68:in `compute_project_path'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer/analyzer.rb:1193:in `each'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer/analyzer.rb:1193:in `group_by'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer/analyzer.rb:1193:in `block in inspect_targets_to_integrate'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/user_interface.rb:64:in `section'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer/analyzer.rb:1188:in `inspect_targets_to_integrate'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer/analyzer.rb:107:in `analyze'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer.rb:422:in `analyze'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer.rb:244:in `block in resolve_dependencies'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/user_interface.rb:64:in `section'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer.rb:243:in `resolve_dependencies'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/installer.rb:162:in `install!'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/command/install.rb:52:in `run'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/claide-1.1.0/lib/claide/command.rb:334:in `run'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/lib/cocoapods/command.rb:52:in `run'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/gems/cocoapods-1.12.1/bin/pod:55:in `<top (required)>'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/bin/pod:25:in `load'
    /opt/homebrew/Cellar/cocoapods/1.12.1/libexec/bin/pod:25:in `<main>'
Error running pod install
Error launching application on iPhone 14 Pro Max.
Exited
```

**Comment by listepo** [comment]
@LouiseHsu @jmagman I think we need flutter command for this

**Pull request merged**
